### PR TITLE
Allow batching for integrated gradient and multithreading for SmoothGrad

### PR DESCRIPTION
**Pull request by sherryy**
We can enable batching if input x and logit y have the same batch size, similar to the answer here: https://stackoverflow.com/questions/51858970/tf-gradients-sums-over-ys-does-it

### CODE REVIEW

File: saliency/base.py [hunk 1]
```diff
@@ -25,18 +26,11 @@ def __init__(self, graph, session, y, x):
...
       y: The output tensor to compute the SaliencyMask against. This tensor
-          should be of size 1.
+          should have the same first dimension (batch size) as x.
       x: The input tensor to compute the SaliencyMask against. The outer
           dimension should be the batch size.
     """
 
-    # y must be of size one, otherwise the gradient we get from tf.gradients
-    # will be summed over all ys.
-    size = 1
```
Comment (nsthorat) [inline saliency/base.py]: can you write a unit test that covers the batch case as you expect? can you explain how you solve this in the description (without having to read stackoverflow)

Comment (sherryy) [reply saliency/base.py]: The existing integrated_gradients_test.py is now testing batching. I'm not batching across different images (there is still only one input image), but batching the x_steps (1000 in the test case) so the 1000 gradients are computed in one session run.

Do you mean I should add comments in the code to explain why batching the x_steps works or just in the description to this PR? If it's the latter, before we thought batching wouldn't work because if x has [batch_size, N] as its shape but y has shape [N] (this is the case because we are feeding y[0] to the constructor in the test), tf.gradient would sum over the first dimension of x. Now allow y to have shape [batch_size, N], we would get a list of gradients of size batch_size. This is what the stackoverflow link is saying too. TensorFlow has to support computing gradient in a batch.

Comment (sherryy) [reply saliency/base.py]: I also tried batching the nsamples when computing the smoothed mask, but that doesn't help much for smoothed integrated gradient because now we are computing nsamples * x_steps masks, which makes the input too big and slow to compute in one session run, so I added multithreading to do concurrent session run to compute smoothed masks.

Comment (nsthorat) [reply saliency/base.py]: Ahhh got it, apologies for misunderstanding!

Comment (sherryy) [reply saliency/base.py]: No worries at all! I was also very confused. Mind approving the pool request?

File: saliency/base.py [hunk 2]
```diff
@@ -53,7 +47,7 @@ def GetMask(self, x_value, feed_dict={}):
 
   def GetSmoothedMask(
       self, x_value, feed_dict={}, stdev_spread=.15, nsamples=25,
-      magnitude=True, **kwargs):
+      magnitude=True, num_threads=1, **kwargs):
```
Comment (dsmilkov) [inline saliency/base.py]: should `num_threads` be -1,so you get the fastest behavior by default?
